core/src/test/scala/quasar/destination/gbq/GBQDestinationSpec.scala

```diff
@@ -0,0 +1,265 @@
+import org.http4s.client.asynchttpclient.AsyncHttpClient
+import org.http4s.client.Client
+import org.http4s.{
+  AuthScheme, 
+  Credentials, 
+  EntityEncoder, 
+  MediaType,
+  Method, 
+  Request, 
+  Status,
+  Uri}
+import org.http4s.headers.Authorization
+import org.http4s.headers.`Content-Type`
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.ExecutionContext.Implicits.global
+
+import scala.util.Right
+import scalaz.{-\/,\/-}
+
+import shims._
+
+object GBQDestinationSpec extends EffectfulQSpec[IO] {
+  sequential
+
+  implicit val timer: Timer[IO] = IO.timer(ExecutionContext.global)
+
+  val testProject = "travis-ci-reform-test-proj"
+  val testDataset = "testdataset"
+  val authCfgPath = Paths.get(getClass.getClassLoader.getResource("gbqAuthFile.json").toURI)
+  val authCfgString: String = new String(Files.readAllBytes(authCfgPath), UTF_8)
+  val cfg = config(authCfg = Some(authCfgString), Some(testProject), Some(testDataset))
+  val resource = mkClient[IO]
+
+  "csv link" should {
+    "reject empty paths with NotAResource" >>* {
+      csv(cfg) { sink =>
+        val p = ResourcePath.root()
+        val r = sink.run(p, List(TableColumn("a", ColumnType.Boolean)), Stream.empty).compile.drain
+
+        MRE.attempt(r).map(_ must beLike {
+          case -\/(ResourceError.NotAResource(p2)) => p2 must_=== p
+        })
+      }
+    }
+
+    "successfully upload table" >>* {
+      val dst = ResourcePath.root() / ResourceName("foo") / ResourceName("bar.csv")
+      val data = Stream("col1,col2\r\nstuff,true\r\n").through(text.utf8Encode)
+
+      csv(cfg) { sink =>
+          val r = sink.run(dst, List(TableColumn("a", ColumnType.String), TableColumn("b", ColumnType.Boolean)), data).compile.drain
+          //TODO: give BigQuery to make it aware of newly pushed table, is there something else we can do?
+          java.lang.Thread.sleep(5000)
+          MRE.attempt(r).map(_ must beLike {
+            case \/-(value) => value must_===(())
+          })
+      }
+    }
+
+    "successfully check dataset was created" >>* {
+      //val resource = mkClient[IO]
+
+      for {
+        accessToken <- GBQAccessToken.token[IO](authCfgString.getBytes("UTF-8"))
+        auth = Authorization(Credentials.Token(AuthScheme.Bearer, accessToken.getTokenValue))
+        req = mkGeneralRequest[String](
+          auth, 
+          "",
+          s"https://content-bigquery.googleapis.com/bigquery/v2/projects/${testProject}/datasets",
+          Method.GET,
+          `Content-Type`(MediaType.application.json)) 
+          resp <- resource.use(client => mkRequest(client, req))
+        containsDatasets <- resp match {
+          case Right(value) =>
+            value.body.compile.toVector.map(v => {
+              val t = v.map(_.toChar).mkString
+              t.contains(testDataset)
+            })
+          case Left(value) => IO{ false }
+        }
+      } yield {
+        containsDatasets must beTrue
```

Comment:
Actually, there are a lot of these cases where you have effectful matchers which are not being wrapped. I won't comment on them all but they do all need to be fixed.